### DISCUSSION TITLE
Automatic lipsync (request)

### DISCUSSION THREAD

**Discussion by butternutbones** [keiwando/toonsquid#538] [Ideas]
I know this is probably a long shot, but incorporating some type of automatic lip sync, similar to (or perhaps somehow using) Rhubarb lip sync, would elevate this app to one of the most powerful animations programs out there. It would be so amazing if we could just draw a few mouth shapes and then use the speech-to-text (or rather speech-to-viseme) of Rhubarb lip sync or similar to auto-generate the lip sync keyframes. 

Toonsquid is already the only iPad animation app (I think?) that has symbols, and I don't know how it's not more popular because of that. It's like the best parts of Flash/Adobe Animate mixed with Procreate. Adding automated lipsync would make it a killer app, IMO.

**Comment by Cat-Butler** [comment]
+1, I'd be delighted if ToonSquid could import one of the Rhubarb output formats (.dat, .json, .tsv, or .xml) to auto-keyframe frame markers.
My personal preference would be .dat, because I also occasionally use Moho, but I'd be happy with any of the other formats.

**Comment by invishane** [comment]
I emailed Keiwando about this exact thing, man it would be amazing to have Auto LipSync. Right now, i used character animator and export a (huge) mov file of ONLY the mouth and add it to the hierarchy where the mouth goes. It works great, the only issue is the export when finished with a project takes FOREVER. So one mistake and i go ….”UUUUUGH”

**Comment by DinoAwakening** [comment]
Not to be a buzzkill, but why you guys would need something like that? I mean, aside of it being utterly impossible in most cases, it still needs all your participation making all those mouth your animation is gonna use, since it isn’t an AI  figuring out your style by looking at the picture, like, it’s not automatic. At this point just use symbols, asking animation to be made by itself takes the fun of making it away. 🦖

**Reply by Cat-Butler** [reply]
Hey @DinoAwakening, I suspect you're coming from a frame-by-frame Disney-style animation viewpoint. And that's a valid question from an animator.
That's like saying "why do you want rigging features, that's taking away from the fun of drawing every frame individually?".
And that's a valid mindset.

However, not all of us have the same mindset.
In an animation context, I don't see myself as primarily an artist, I'm primarily a storyteller.
I want rigging so I can create a character view once and then "puppet" it as needed to tell the story. ToonSquid is one of the very few iPad apps to enable that sort of approach.

I don't find aligning pre-set frames from a mouth Symbol to voice audio to be a fun or creative activity. It's a tedious repetitive activity that the computer/tablet should do: Frame 3, use this frame in the mouth Symbol. Frame 7, use this frame in the mouth Symbol. Frame 11, use this frame in the mouth Symbol. Bleh. 😝 

Creating the original mouth shapes for the Symbol, yes, that's a fun creative endeavour, maybe more than one set for different moods and emotions. That's for the animator to do manually. Auto-lipsync doesn't do that. 
Instead, I want to feed it an audio file, a text file of the dialogue, and it can darn well align those mouth shape keyframes for me, using my carefully crafted mouth Symbol(s). Maybe I'll edit them further manually afterwards.

I'll note that personally I usually prefer Hanna-Barbera, Aardman, Astley Baker Davies, and similar rigged and stylised animation with good storylines or humour  (I can still get a chuckle when occasionally watching [The Big Knights](https://www.youtube.com/@thebigknights4465/videos)), far more than overdrawn frame-by-frame Disney-like styles and overly dramatic action.
There are exceptions, of course, but that's my general preference, probably influenced by childhood cartoon viewing (I'm old).

And yes, it's do-able. Is it easy to add into ToonSquid without major code changes? I don't know. It might be difficult. It might be darn near impossible. But I think it's worth asking.

It's why auto-lipsync via apps (Papagayo, Rhubarb), extensions, and built-in functionality are popular for users of desktop rigging-style character animation apps like Adobe Animate, Moho, Toon Boom Harmony, CelAction 2d, Cartoon Animator, Synfig, and for some uses of OpenToonz, Clip Studio Paint, Blender, and even for character animation in After Effects.

**Comment by NavodPeiris** [comment]
@butternutbones you might be interested in something like this: https://github.com/NavodPeiris/agentic-avatars
It does AI avatar lipsync directly on browser

**Comment by dougalder** [comment]
I made a simple lip sync web app. This might be useful until ToonSquid adds this feature request. 

[Lip Sync Generator](https://dougalder.github.io/lip-sync-generator/lip-sync-generator.html). Free, open source, no downloads, private. [Project Home](https://github.com/dougalder/lip-sync-generator). 15 mouth styles in Front, 3/4s and Profile views. 7 Export types: Transparent GIF, Animated PNG, PNG sequence, Final Cut Pro xml, mouth overlay video, green screen and Chart and Exposure sheet and Sprite Sheet and Atlas. You can add a puppet still or video in the background and export with that, or simply export the mouth and use that in ToonSquid. 

Import the mouths and the voice track to ToonSquid, drag them into the timeline above your puppet, and resize and position them there. Keep the animated mouth attached to the head of your character so the two travel together, which is worth setting up before you animate anything -- use the transform hierarchy — the button in the timeline's bottom toolbar. Drag the mouth layer onto the body layer to make it a child, and scaling, rotating or moving the body does the same to the mouth.  Position the voice track in the appropriate place below the mouth.

<img width="640" height="480" alt="lip-sync-generator-demo-short" src="https://github.com/user-attachments/assets/4e995163-a2d8-43d1-a3db-1e44c85e8bf9" />
